drivers/adc/ad7091r8/ad7091r8.c

```diff
@@ -0,0 +1,578 @@
+	}
+}
+
+/***************************************************************************//**
+ * Initializes the communication peripheral and the initial Values for
+ * AD7092R-8 Board.
+ * @param device     - The device structure.
+ * @param init_param - The structure that contains the device initial
+ * 		       parameters.
+ *
+ * @return ret - The result of the initialization procedure.
+ *               Example: -1 - SPI peripheral was not initialized or the
+ *                             device is not present.
+ *                         0 - SPI peripheral was initialized and the
+ *                             device is present.
+*******************************************************************************/
+int ad7091r8_init(struct ad7091r8_dev **device,
+		  struct ad7091r8_init_param *init_param)
+{
+	struct ad7091r8_dev *dev;
+	int ret;
+
+	if (!device)
+		return -EINVAL;
+
+	dev = (struct ad7091r8_dev *)no_os_calloc(1, sizeof(*dev));
+	if (!dev)
+		return -ENOMEM;
+
+	ret = no_os_spi_init(&dev->spi_desc, init_param->spi_init);
+	if (ret) {
+		printf("Failed to init SPI device: %d\n\r", ret);
+		goto err_release_spi;
+	}
+
+	dev->device_id = init_param->device_id;
+
+	ret = no_os_gpio_get(&dev->gpio_convst, init_param->gpio_convst);
+	if (ret) {
+		printf("Failed to get CONVST GPIO\n\r");
+		goto err_release_convst;
+	}
+
+	ret = no_os_gpio_direction_output(dev->gpio_convst, NO_OS_GPIO_HIGH);
+	if (ret) {
+		printf("Failed to set CONVST GPIO output\n\r");
+		goto err_release_convst;
+	}
+
+	ret = no_os_gpio_get_optional(&dev->gpio_reset, init_param->gpio_reset);
+	if (!ret && dev->gpio_reset) {
+		ret = no_os_gpio_direction_output(dev->gpio_reset,
+						  NO_OS_GPIO_HIGH);
+		if (ret)
+			printf("WARNING: Failed to set RESET GPIO output\n\r");
+	}
+
+	ret = ad7091r8_reset(dev, !dev->gpio_reset);
+	if (ret)
+		printf("WARNING: Failed to RESET on power up\n\r");
+
+	/* Use external vref or enable internal vref */
+	dev->vref_mv = init_param->vref_mv;
+	if (!dev->vref_mv) {
+		dev->vref_mv = 2500;
+		ad7091r8_set_sleep_mode(dev, AD7091R8_SLEEP_MODE_1);
+		if (ret)
+			printf("WARNING: Failed to enable internal vref\n\r");
+	}
+
+	/* Device powers-up in normal mode */
+	*device = dev;
+
+	return 0;
+
+err_release_convst:
+	if(no_os_gpio_remove(dev->gpio_convst))
+		printf("Failed to realease CONVST GPIO\n\r");
+
+err_release_spi:
+	if(no_os_spi_remove(dev->spi_desc))
+		printf("Failed to realease SPI descriptor\n\r");
+
+	no_os_free(dev);
+	return ret;
+}
+
+/***************************************************************************//**
+ * @brief Free the resources allocated by ad7091r8_init().
+ *
+ * @param dev - The device structure.
+ *
+ * @return ret - The result of the remove procedure.
+*******************************************************************************/
+int ad7091r8_remove(struct ad7091r8_dev *dev)
+{
+	int ret;
+
+	ret = no_os_gpio_remove(dev->gpio_reset);
+	if (ret)
+		printf("Failed to realease RESET GPIO");
```

Comment:
Functions from no_os_print_log.h would be more appropriate for printing messages (e.g., pr_info, pr_err, ...).